### ISSUE TITLE
How to build a gb project with vendored gb projects?

### ISSUE THREAD

**Issue by goloroden** [constabulary/gb#443]
I am doing my first steps in Go, and I have quite quickly stumbled upon the problem of vendoring. I was pointed to `gb` and I really like the approach this tool is taking, especially also that it works on a project-base instead of a workspace-base.

So far, I have written a library project and I can build it using `gb build`. This works fine :-)

Now I have pushed this library to GitHub, and started an actual program, in which I would like to use my library. So I have run:

    $ gb vendor fetch github.com/thenativeweb/buntstift-go

This downloaded everything and again, so far, so good. But if I now try to compile my program, I get an error message telling me:

> FATAL: command "build" failed: failed to resolve import path "main": no buildable Go source files in /Users/golo/Projekte/thenativeweb/hello-world/vendor/src/github.com/thenativeweb/buntstift-go

As it seems, `gb` does not resolve a library's file and folder structure, if this library itself uses `gb`. Am I missing something, or is this actually something not working?

**Comment by oleksandr** [comment]
+1

**Comment by davecheney** [comment]
Thank you for logging this issue. I know that people want to use gb vendor to vendor other gb projects. This is a known issue, and one of the oldest #49. 

There is logic in gb-vendor to support vendoring and excluding a path (ie, the src/github.com/blah/blah) but it is not hooked up and will probably need a version number bump in the manifest.

This is something I am aware of, but apart from the roadmap in the README file, I cannot tell you when it will be addressed.

**Comment by metral** [comment]
+1

**Comment by davecheney** [comment]
Hello,

Thank you for your plus ones, while they aren't that useful, imdo.understand that people want to use gb for libraries. 

To set expectations, this is not on the roadmap at this time. It's not that it's not important, its just not more important than the things on the roadmap.

**Comment by metral** [comment]
Sorry for the lack of info on the +1 - yes, I was wanting to be able to use gb for libraries with other gb projects. Thanks for the prompt response

**Comment by seh** [comment]
Is the advice, then, that if you're writing a library, _don't_ set it up with gb's directory structure, and don't use gb to build it? That is, use gb only for the "root vertices" in the dependency graph?

**Comment by davecheney** [comment]
Nope, you'll just have to vendor your code manually. If you're fine with
that, then it should just work (tm)

**Comment by seh** [comment]
But will using, say, a Git submodule to include the whole repository for the library provide a directory structure that gb will accept under the _vendor_ directory? I take it that if the library itself has vendored dependencies, we'd have to vendor those manually too, so that we lose the benefit of having the library in its own source repository.

**Comment by davecheney** [comment]
Should work, I pitched submodules was back in the beginning for managing
$PROJECT/vendors/src but people weren't buying it, so I wrote gb vendor.

Gb doesn't care how the source gets into that directory, just that it is
there in the right directory structure.

**Comment by seh** [comment]
Is it the case, though, that for a gb-built library with dependencies included in a gb-built application, the application would need to vendor not only that library, but all of the library's dependencies at the "top level"?

**Comment by davecheney** [comment]
Yes

On Sun, Jan 17, 2016 at 12:15 AM, Steven E. Harris <notifications@github.com

**Comment by oleksandr** [comment]
This is exactly the case we're suffering at the moment. The complete product is a set of ~20 independent services (gb projects, developed by different engineers) that have a shared library (a gb-project), which contain not only shared code but also optional tools (cmds). Unfortunately, we have to mirror deps from shared project in every of those 20 projects.

As as soon as someone adds a useful function/sub-package to shared library that has its own deps - we can't simply use it from those 20 projects. First we have to vendor the dependency... Can live with that - but stumble upon each time :)

**Comment by seh** [comment]
And how do you vendor those dependencies into each of those 20 other projects?

**Comment by metral** [comment]
I personally wrote this shim to get around the vendoring of a gb library project in other gb app projects : https://github.com/metral/basics/blob/master/gb-manual-vendor-update.sh

From a gb app project:
1) pull the repo/branch of the gb lib proj
2) make a copy of the vendor manifest for the deps of the gb app to remember what the projects vendoring looks like
3) copy the gb lib proj src/ & vendor/src into the gb app vendor/src/
4) use the gb app's vendor manifest to restore its own deps

it's not pretty, but it works

**Comment by oleksandr** [comment]
@seh, we have to keep shared's vendors in each prj's vendors although they are not used directly, only via shared lib. The layout is something like:

```
/product-dir/prj1/src/prj1/
/product-dir/prj1/vendor/
/product-dir/prj1/vendor/src/shared -> sym link to /product-dir/shared/src/shared
/product-dir/prj2/src/prj2
/product-dir/prj2/vendor/
/product-dir/prj2/vendor/src/shared -> sym link to /product-dir/shared/src/shared
...
/product-dir/shared/src/shared
/product-dir/shared/vendor/
```

The setup is scripted using invoked (e.g. `inv setup`) which is also kind of ugly, but works.

**Comment by oleksandr** [comment]
@metral nice walk around :-)

**Comment by seh** [comment]
Thanks, @oleksandr. That makes sense.
You mentioned symbolic links. Do you ever use Git submodules in their place?
I'd also like to see your [_invoke_](http://www.pyinvoke.org) configuration for this, if you're willing to share.

**Comment by jmouradi** [comment]
Sorry to provide a late remark on such an old issue, but I wanted to address the following comment, especially since the first occurrence of this issue was over two years ago: "It's not that it's not important, its just not more important than the things on the roadmap."

There may be important features on the roadmap, but people for whom this is a dealbreaker are simply moving away from your tool. We greatly enjoyed having a project-based dependency management solution that allowed us to place code wherever we wanted, but we use common libraries across multiple projects. As such, to make dependency management a homogeneous process, we have moved to [Glide](https://github.com/Masterminds/glide).
